### PR TITLE
feat: Add Qwen3 MoE custom implementation

### PR DESCRIPTION
**Pull request by hemildesai**

### CODE REVIEW

**Comment by hemildesai** [comment]
/ok to test [7f9ba2b](https://github.com/NVIDIA-NeMo/Automodel/pull/507/commits/7f9ba2bd92050e0bdd35543e64a333a184f46a4c)

**Comment by hemildesai** [comment]
/ok to test [8d04431](https://github.com/NVIDIA-NeMo/Automodel/pull/507/commits/8d044318f4ae26e25da7cbb575e678d53f858109)

**Comment by hemildesai** [comment]
/ok to test [de1322c](https://github.com/NVIDIA-NeMo/Automodel/pull/507/commits/de1322c3ed31a1019a2060c27bdfe5c33872b9ac)

**Comment by hemildesai** [comment]
/ok to test [4166baf](https://github.com/NVIDIA-NeMo/Automodel/pull/507/commits/4166bafafffd45204c0513dd4a9ef96727b1b443)

**Comment by hemildesai** [comment]
/ok to test [f8e3b4f](https://github.com/NVIDIA-NeMo/Automodel/pull/507/commits/f8e3b4f87f2031f37322546eabe7d503dbe564f5)

File: nemo_automodel/components/moe/layers.py [hunk 1]
```diff
@@ -495,6 +496,8 @@ def forward(
                 output2 = self._apply_bias(output2, down_bias, tokens_per_expert, permuted_probs)
         else:
             output2 = torch.zeros((1, x.size(-1)), dtype=x.dtype, device=x.device)
+            output2 = output2 * permuted_probs
+            output2 = output2.to(x.dtype)
```
Comment (adil-a) [inline nemo_automodel/components/moe/layers.py:500]: why do we need this now? have you tested it doesn't cause issues for the other moe models?

File: nemo_automodel/components/moe/layers.py [hunk 2]
```diff
@@ -660,12 +670,12 @@ def forward(
...
-                denom_w = weights.sum(dim=-1, keepdim=True) + 1e-20
-                denom_s = original_scores.sum(dim=-1, keepdim=True) + 1e-20
-                weights = weights / denom_w
-                original_scores = original_scores / denom_s
+        if self.norm_topk_prob and self.topk > 1:
+            # Use out-of-place division to keep autograd versions intact.
+            denom_w = weights.sum(dim=-1, keepdim=True) + 1e-20
+            denom_s = original_scores.sum(dim=-1, keepdim=True) + 1e-20
+            weights = weights / denom_w
+            original_scores = original_scores / denom_s
```
Comment (adil-a) [inline nemo_automodel/components/moe/layers.py:678]: seems like this will now be applied in the softmax case too after this change. Is this intended?

File: nemo_automodel/recipes/llm/train_ft.py [hunk 3]
```diff
@@ -79,9 +79,17 @@
...
 # ---------------------------
 #  Stateless helper functions
 # ---------------------------
+def _get_model_name(cfg_model):
+    if cfg_model.get("pretrained_model_name_or_path", None) is not None:
+        return cfg_model.pretrained_model_name_or_path
+    elif cfg_model.get("config", None) is not None:
+        return cfg_model.config.get("pretrained_model_name_or_path", None)
+    else:
+        return None
```
Comment (adil-a) [inline nemo_automodel/recipes/llm/train_ft.py:92]: lets add this functionality for VLM recipe too?

**Review by adil-a — COMMENTED** [review]
left some comments, otherwise lgtm! thank you :)

File: nemo_automodel/components/moe/layers.py (hunk 2 above)
Comment (hemildesai) [reply nemo_automodel/components/moe/layers.py:678]: Yes this was because Qwen3 uses this normalization as well.

File: nemo_automodel/recipes/llm/train_ft.py (hunk 3 above)
Comment (hemildesai) [reply nemo_automodel/recipes/llm/train_ft.py:92]: I can follow up in a separate PR for this

File: nemo_automodel/components/moe/layers.py (hunk 1 above)
Comment (hemildesai) [reply nemo_automodel/components/moe/layers.py:500]: Talked offline, will follow up in https://github.com/NVIDIA-NeMo/Automodel/pull/532

**Review by adil-a — APPROVED** [review]
lgtm, thank you
